File: modules/widget-bridge/ios/WidgetBridgeModule.swift
import ExpoModulesCore
#if canImport(WidgetKit)
import WidgetKit
#endif

public class WidgetBridgeModule: Module {
  private let suiteName = "group.com.dayink.app"
  private let snapshotKey = "dailySnapshot"
  private let levelKey = "activeLevel"
  private let shownKey = "shownYearByWordId"

  public func definition() -> ModuleDefinition {
    Name("WidgetBridge")

    AsyncFunction("syncWidgetState") { (snapshot: [String: String]?, level: String?) in
      let defaults = UserDefaults(suiteName: self.suiteName) ?? .standard
      if let snapshot,
         let data = try? JSONSerialization.data(withJSONObject: snapshot),
         let json = String(data: data, encoding: .utf8) {
        defaults.set(json, forKey: self.snapshotKey)
      }
      if let level {
        defaults.set(level, forKey: self.levelKey)
      }
      defaults.synchronize()
    }

    AsyncFunction("setDailySnapshot") { (snapshot: [String: String]) in
      let defaults = UserDefaults(suiteName: self.suiteName) ?? .standard
      if let data = try? JSONSerialization.data(withJSONObject: snapshot),
         let json = String(data: data, encoding: .utf8) {
        defaults.set(json, forKey: self.snapshotKey)
        defaults.synchronize()
      }
    }

    AsyncFunction("setActiveLevel") { (level: String) in
      let defaults = UserDefaults(suiteName: self.suiteName) ?? .standard
      defaults.set(level, forKey: self.levelKey)
      defaults.synchronize()
    }

    AsyncFunction("setShownYears") { (json: String) in
      let defaults = UserDefaults(suiteName: self.suiteName) ?? .standard
      defaults.set(json, forKey: self.shownKey)
      defaults.synchronize()
    }

    AsyncFunction("getDailySnapshot") { () -> [String: String]? in
      let defaults = UserDefaults(suiteName: self.suiteName) ?? .standard
      guard let json = defaults.string(forKey: self.snapshotKey),
            let data = json.data(using: .utf8),
            let object = try? JSONSerialization.jsonObject(with: data) as? [String: String] else {
        return nil
      }
      return object
    }

    AsyncFunction("reloadWidgets") {
      #if canImport(WidgetKit)
      if #available(iOS 14.0, *) {
        WidgetCenter.shared.reloadTimelines(ofKind: "DayinkWidget")
        WidgetCenter.shared.reloadAllTimelines()
      }
      #endif
    }
  }
}

File: modules/widget-bridge/android/src/main/java/expo/modules/widgetbridge/WidgetBridgeModule.kt
package expo.modules.widgetbridge

import android.content.Context
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import org.json.JSONObject

class WidgetBridgeModule : Module() {
  private val prefsName = "dayink_widget"
  private val snapshotKey = "dailySnapshot"
  private val levelKey = "activeLevel"
  private val shownKey = "shownYearByWordId"

  override fun definition() = ModuleDefinition {
    Name("WidgetBridge")

    AsyncFunction("syncWidgetState") { snapshot: Map<String, String>?, level: String? ->
      val context = appContext.reactContext ?: return@AsyncFunction
      val editor = context.getSharedPreferences(prefsName, Context.MODE_PRIVATE).edit()
      if (snapshot != null) {
        val json = JSONObject()
        for ((key, value) in snapshot) {
          json.put(key, value)
        }
        editor.putString(snapshotKey, json.toString())
      }
      if (level != null) {
        editor.putString(levelKey, level)
      }
      editor.commit()
    }

    AsyncFunction("setDailySnapshot") { snapshot: Map<String, String> ->
      val context = appContext.reactContext ?: return@AsyncFunction
      val json = JSONObject()
      for ((key, value) in snapshot) {
        json.put(key, value)
      }
      context.getSharedPreferences(prefsName, Context.MODE_PRIVATE)
        .edit()
        .putString(snapshotKey, json.toString())
        .commit()
    }

    AsyncFunction("setActiveLevel") { level: String ->
      val context = appContext.reactContext ?: return@AsyncFunction
      context.getSharedPreferences(prefsName, Context.MODE_PRIVATE)
        .edit()
        .putString(levelKey, level)
        .commit()
    }

    AsyncFunction("setShownYears") { json: String ->
      val context = appContext.reactContext ?: return@AsyncFunction
      context.getSharedPreferences(prefsName, Context.MODE_PRIVATE)
        .edit()
        .putString(shownKey, json)
        .commit()
    }

    AsyncFunction("getDailySnapshot") {
      val context = appContext.reactContext ?: return@AsyncFunction null
      val raw = context.getSharedPreferences(prefsName, Context.MODE_PRIVATE)
        .getString(snapshotKey, null) ?: return@AsyncFunction null
      val obj = JSONObject(raw)
      val map = mutableMapOf<String, String>()
      val keys = obj.keys()
      while (keys.hasNext()) {
        val key = keys.next()
        map[key] = obj.getString(key)
      }
      map
    }

    AsyncFunction("reloadWidgets") {
      val context = appContext.reactContext ?: return@AsyncFunction
      DayinkWidgetUpdater.requestUpdate(context)
    }
  }
}
